cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -1072,6 +1133,74 @@ private ProgramId validateProgramForRunRecord(String namespaceName, String appNa
   }
 
   /**
+   * Helper method to update log levels for Worker, Flow, or Service.
+   */
+  private void updateLogLevels(ProgramId programId, Map<String, String> logLevels, @Nullable String component)
+    throws Exception {
+    ProgramRuntimeService.RuntimeInfo runtimeInfo = findRuntimeInfo(programId);
+    if (runtimeInfo != null) {
+      TwillController twillController = getTwillController(runtimeInfo);
+      if (component == null) {
+        twillController.updateLogLevels(transformLogLevelsMap(logLevels)).get();
+      } else {
+        twillController.updateLogLevels(component, transformLogLevelsMap(logLevels)).get();
+      }
+    }
+  }
+
+  /**
+   * Helper method to reset log levels for Worker, Flow or Service.
+   */
+  private void resetLogLevels(ProgramId programId, @Nullable String[] loggerNames, @Nullable String component)
+    throws Exception {
+    ProgramRuntimeService.RuntimeInfo runtimeInfo = findRuntimeInfo(programId);
+    if (runtimeInfo != null) {
+      TwillController twillController = getTwillController(runtimeInfo);
+      if (component == null) {
+        if (loggerNames == null || loggerNames.length == 0) {
+          twillController.resetLogLevels().get();
+        } else {
+          twillController.resetLogLevels(loggerNames).get();
+        }
+      } else {
+        if (loggerNames == null || loggerNames.length == 0) {
+          twillController.resetRunnableLogLevels(component).get();
+        } else {
+          twillController.resetRunnableLogLevels(component, loggerNames).get();
+        }
+      }
+    }
+  }
+
+  /**
+   * Helper method to get the {@link TwillController} for the program.
+   */
+  private TwillController getTwillController(ProgramRuntimeService.RuntimeInfo runtimeInfo) throws Exception {
+    ProgramController programController = runtimeInfo.getController();
+    if (!(programController instanceof AbstractTwillProgramController)) {
```

Comment:
I think I had a discussion with @chtyim to cast the ProgramController here so that it is clearer to show that log level change at runtime is only supported at distributed. Will confirm with him tomorrow